source/extensions/filters/network/thrift_proxy/router/router_impl.cc

```diff
@@ -454,8 +463,29 @@ void Router::UpstreamRequest::onRequestStart(bool continue_decoding) {
     parent_.callbacks_->continueDecoding();
   }
 }
+void Router::UpstreamRequest::onRequestComplete() {
+  request_complete_ = true;
+
+  if (parent_.timeout_.count() > 0) {
+    parent_.response_timer_ = parent_.callbacks_->dispatcher().createTimer(
```

Comment:
Oh. That's why time_system_ exists -- you get a SchedulerPtr from TimeSystem and use that to create timers. Those timers will be thread-local and should fix the weird thread safety asserts you got in CI.

Reply:
interesting -- so i obviously modeled this a lot after what's done for http and from what I understand, timers [like these](https://github.com/envoyproxy/envoy/blob/8c6326e893db04aafb959d5bf6d9a367803e54b2/source/common/http/conn_manager_impl.cc#L384) don't _seem_ to be created on the dispatcher thread. how's that work...?